android/quest/src/main/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireDraftDialogViewModel.kt

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2021-2024 Ona Systems, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.smartregister.fhircore.quest.ui.questionnaire
+
+import androidx.lifecycle.ViewModel
+import dagger.hilt.android.lifecycle.HiltViewModel
+import java.util.Date
+import javax.inject.Inject
+import org.hl7.fhir.r4.model.AuditEvent
+import org.hl7.fhir.r4.model.AuditEvent.AuditEventSourceComponent
+import org.hl7.fhir.r4.model.Coding
+import org.hl7.fhir.r4.model.Period
+import org.hl7.fhir.r4.model.QuestionnaireResponse
+import org.hl7.fhir.r4.model.QuestionnaireResponse.QuestionnaireResponseStatus
+import org.hl7.fhir.r4.model.Reference
+import org.hl7.fhir.r4.model.ResourceType
+import org.smartregister.fhircore.engine.configuration.QuestionnaireConfig
+import org.smartregister.fhircore.engine.data.local.DefaultRepository
+import org.smartregister.fhircore.engine.util.SharedPreferenceKey
+import org.smartregister.fhircore.engine.util.SharedPreferencesHelper
+import org.smartregister.fhircore.engine.util.extension.asReference
+import org.smartregister.fhircore.engine.util.extension.extractLogicalIdUuid
+
+@HiltViewModel
+class QuestionnaireDraftDialogViewModel
+@Inject
+constructor(
+  val defaultRepository: DefaultRepository,
+  val sharedPreferencesHelper: SharedPreferencesHelper,
+) : ViewModel() {
+
+  private val practitionerId: String? by lazy {
+    sharedPreferencesHelper
+      .read(SharedPreferenceKey.PRACTITIONER_ID.name, null)
+      ?.extractLogicalIdUuid()
+  }
+
+  suspend fun deleteDraft(questionnaireConfig: QuestionnaireConfig?) {
+    if (
+      questionnaireConfig == null ||
+        questionnaireConfig.resourceIdentifier.isNullOrBlank() ||
+        questionnaireConfig.resourceType == null
+    ) {
+      return
+    }
+
+    val questionnaireResponse =
+      defaultRepository.searchQuestionnaireResponse(
+        resourceId = questionnaireConfig.resourceIdentifier!!,
+        resourceType = questionnaireConfig.resourceType!!,
+        questionnaireId = questionnaireConfig.id,
+        encounterId = null,
+        questionnaireResponseStatus = QuestionnaireResponseStatus.INPROGRESS.toCode(),
+      )
+
+    if (questionnaireResponse != null) {
+      questionnaireResponse.status = QuestionnaireResponseStatus.STOPPED
+      defaultRepository.update(questionnaireResponse)
+      defaultRepository.addOrUpdate(
```

Comment:
In the unlikely event that one passes and the other fails, might be best to wrap this with a DB transaction.

Reply:
We'd probably want to create a new method in defaultRespository for this and then within it use the txn block
```kotlin
fhirEngine.withTransaction {  
  ...
}